### PR TITLE
[rl] Fix batch invariant logprob calculation  by forcing vllm use trainer's function

### PR DESCRIPTION
**Pull request by wwwjn**
Rational: The logp calculation should be more transparent to us

### CODE REVIEW

File: torchtitan/experiments/rl/actors/generator.py [hunk 1]
```diff
@@ -98,6 +98,38 @@ def _prepare_generation_request_metrics(
...
+
+    from torchtitan.experiments.rl.actors.trainer import (
+        compute_logprobs as trainer_compute_logprobs_fn,
+    )
+
+    def generator_compute_token_logprobs(
+        logits: torch.Tensor, token_ids: torch.Tensor
+    ) -> torch.Tensor:
+        # Broadcast 2D logits over token_ids' num_logprobs columns so the
+        # trainer's compute_logprobs (one token per position) works unchanged.
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/generator.py]: say more. I'm confused by when we need to `expand` anything. The docstring of https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/actors/trainer.py#L45 is not great, either. But at least I could see the the BSV stands for bsz, seqlen, vocab size.

Comment (wwwjn) [reply torchtitan/experiments/rl/actors/generator.py]: Thanks for pushing back, this actually help me find a important bug - expand only satisfy the shape requirement, but the meaning of each element in the tensor will be wrong.

I updated it to be more explict, and notice the important different between vllm's compute_logprob with ours:
- vllm computes logprobs for K token_ids at the same position. This is because sometime user need to return more logprobs than just the sampled token id, user might want top-x logprobs at that position (So K=1+x. 1 is the sampled token's logprob). https://docs.vllm.ai/en/latest/api/vllm/sampling_params/#vllm.sampling_params.SamplingParams (search for `logprobs: int | None = None`)

File: torchtitan/experiments/rl/actors/generator.py [hunk 2]
```diff
@@ -426,7 +477,10 @@ async def generate(
                 n=1,  # group_size pre-expands prompts; the RL loop always samples n=1
                 stop_token_ids=self._stop_token_ids or None,
                 seed=self.config.debug.seed,
-                logprobs=1,
+                # NOTE: logprobs=0 -> vLLM returns the sampled token's logprob
+                # only (no top-k alternatives), which is exactly what RL compares
+                # against the trainer's recompute.
+                logprobs=0,
```
Comment (liangel-02) [inline torchtitan/experiments/rl/actors/generator.py]: clarifying question for my understanding: without changing this, even with the patch `_force_logprobs_fn_for_batch_invariance`, would vllm compute K=1+x=2 token ids instead of 1?

File: torchtitan/experiments/rl/actors/generator.py [hunk 3]
```diff
@@ -98,6 +98,53 @@ def _prepare_generation_request_metrics(
...
+        """
+        # Map the N positions to a single sequence (B=1, S=N) and reuse the
+        # trainer's compute_logprobs for each of the K target columns: it scores
+        # one token per position, so K calls are needed (K=1 for our logprobs=0
+        # requests; vLLM's warmup probes K>1).
+        logits = logits.unsqueeze(0)  # [1, N, V]  (B=1, S=N)
+        token_ids = token_ids.to(torch.int64)
+        per_column = [
+            compute_logprobs(logits, token_ids[:, k].unsqueeze(0))  # [1, N]
+            for k in range(token_ids.shape[1])
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/generator.py:149]: since you do `SamplingParams(logprobs=0)`, can we assert `k==1` and simplify the code and doc here.

File: torchtitan/experiments/rl/actors/generator.py (hunk 2 above)
Comment (wwwjn) [reply torchtitan/experiments/rl/actors/generator.py]: yes this is the correct understanding. Previously we are using [0] to slice only the first token (actual sampled token) logprob

File: torchtitan/experiments/rl/actors/generator.py (hunk 3 above)
Comment (wwwjn) [reply torchtitan/experiments/rl/actors/generator.py:149]: Actually we can not assert k==1 here, because vllm warm up runs will also call this function, and pass K > 1 to this function. As a generalized function, this function should support any K (that's why we need the loop here)

So I added doc around `SamplingParams(logprobs=0)` instead

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/experiments/rl/actors/generator.py (hunk 3 above)
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/generator.py:149]: actually I'm not comfortable with this implementation, without knowing what it means for `logprobs>=1`, please educate me

Comment (wwwjn) [reply torchtitan/experiments/rl/actors/generator.py:149]: Sure let me explain: logprob >=1 means: besides the sampled token_id at position i, I also want to know other top-x token's logprobs at position i. Here's the top-x token_ids (stored in the token_ids tensor, shape is [N, K], N=total number of tokens, K = x+1 ) , please give me all x+1 token's logprob. 

And the loop over K is because trainer's compute logprob function only takes one token as labels once at a time: https://github.com/pytorch/torchtitan/blob/bb453e0ca68069483207f28a881ca56c5d57e23c/torchtitan/experiments/rl/actors/trainer.py#L45

Comment (wwwjn) [reply torchtitan/experiments/rl/actors/generator.py:149]: Here's how vllm construct the input to vllm's compute_logprob function:

https://github.com/vllm-project/vllm/blob/e0871ad2259768add6dc43e2972bd364d0d13086/vllm/v1/worker/gpu/sample/logprob.py#L116-L118

- The `sampled_token_ids` are already materialized, and also the `logits` corresponding to that who sequence are also materialized. So the `K = 1 + num_logprobs` tokens at position `i` are: the sampled token, and the top-`num_logprobs` highest-logit tokens of position i's own next-token distribution (torch.topk over the vocab V).
- The topk at position i is over logits[i], which is conditioned on the actually realized tokens at positions 0..i-1 
- There's also a non-topk path ([here](https://github.com/vllm-project/vllm/blob/e0871ad2259768add6dc43e2972bd364d0d13086/vllm/v1/worker/gpu/sample/logprob.py#L120C5-L127C1)): if a user sets `SamplingParams.logprob_token_ids`, the user passes exact token ids to score (not topk). 
- And this is similiar to OpenAI-style logprobs API (which also take a top_k_logprobs) . A real use case why vllm expose `logprobs` parameter is (from claude): confidence display, debugging, reranking/best-of-n scoring
  - Confidence display: people want to see which token the model unsure, and probably want to gather a overall confidence score to this whole answer (more like a research topic)
   - debugging could also be a valid reason - eg, people want to see if top seveal tokens have simliar logprobs which means the sampling strategy is very import to decide the actual generation

Comment (tianyu-l) [reply torchtitan/experiments/rl/actors/generator.py:149]: awesome, thanks a lot!

Please add a NOTE that for each element in token_ids, the computation of logprob is independent. We are compute in a column-wise fashion simply because the existing `compute_logprobs` function in torchtitan only takes one column at a time.

**Review by liangel-02 — APPROVED** [review]
